presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -15,101 +15,125 @@
+        checkArgument(targetQueueSize >= 1, "targetQueueSize must be at least 1");
+        this.targetQueueSize = targetQueueSize;
+        this.elements = new ArrayDeque<>(targetQueueSize * 2);
+        this.executor = requireNonNull(executor);
     }
 
-    /**
-     * Gets a batch of elements from the queue.  If queue is not empty, a
-     * completed future containing the queued elements is returned.  If the
-     * queue is empty, an incomplete future is returned that completes when
-     * an element is added.
-     * <p>
-     * It is possible that an empty list will be returned as another reader
-     * may have consumed all of the data.  A caller should read until, the
-     * {@code isFinished} method returns {@code true}.
-     *
-     * @param maxSize the maximum number of elements to return.
-     */
-    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    public synchronized int size()
     {
-        return notEmptyFuture.get().thenApply(x -> getBatch(maxSize));
+        return elements.size();
     }
 
-    private List<T> getBatch(int maxSize)
+    public synchronized boolean isFinished()
     {
-        // take up to maxSize elements from the queue
-        List<Object> elements = new ArrayList<>(maxSize);
-        queue.drainTo(elements, maxSize);
-
-        // if the queue is empty and the current future is finished, create a new one so
-        // a new readers can be notified when the queue has elements to read
-        if (queue.isEmpty() && !finished.get()) {
-            CompletableFuture<?> future = notEmptyFuture.get();
-            if (future.isDone()) {
-                notEmptyFuture.compareAndSet(future, new CompletableFuture<>());
-            }
+        return finishing && elements.size() == 0;
+    }
+
+    public synchronized void finish()
+    {
+        if (finishing) {
+            return;
+        }
+        finishing = true;
+
+        if (elements.size() == 0) {
+            completeAsync(executor, notEmptySignal);
+            notEmptySignal = new CompletableFuture<>();
         }
+        else if (elements.size() >= targetQueueSize) {
+            completeAsync(executor, notFullSignal);
+            notFullSignal = new CompletableFuture<>();
+        }
+    }
+
+    public synchronized CompletableFuture<?> offer(T element)
+    {
+        requireNonNull(element);
 
-        // if someone added an element while we were updating the state, complete the current future
-        if (!queue.isEmpty() || finished.get()) {
-            notEmptyFuture.get().complete(null);
+        if (finishing) {
+            return CompletableFuture.completedFuture(null);
         }
+        elements.add(element);
+        int newSize = elements.size();
+        if (newSize == 1) {
+            completeAsync(executor, notEmptySignal);
+            notEmptySignal = new CompletableFuture<>();
+        }
+        return newSize < targetQueueSize ? CompletableFuture.completedFuture(null) : notFullSignal;
```

Comment:
This might be slightly easier to read as:

```java
if (newSize >= targetQueueSize) {
    return notFullSignal;
}

return CompletableFuture.completedFuture(null);
```